{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_output_dir = \"/Users/isammohdibrahim/Documents/Dev/running-with-oliver-cromwell/output\"\n",
    "input_file = \"/Users/isammohdibrahim/Documents/Dev/running-with-oliver-cromwell/subsample.csv\"\n",
    "\n",
    "#this is uaully a reference to the accession or source of the data\n",
    "groupname = \"crc_sub1\" \n",
    "\n",
    "ref_version = \"b37\" # or hg38 \n",
    "#this is location of the reference files on ARASHI. Don't change unless intending to use the pipeline with own set of references not defined here\n",
    "ref_dir = f\"/home/imibrahim/refs\"\n",
    "\n",
    "# this notebook expects the aligned bam files for tumor and normal samples to be separated\n",
    "aligned_bam_dir = \"/home/imibrahim/crc_sub1/bam\"\n",
    "n_dir = f'{aligned_bam_dir}/{ref_version}-n'\n",
    "t_dir = f'{aligned_bam_dir}/{ref_version}-t'\n",
    "\n",
    "send_email_for_all = True\n",
    "email = \"isambmi@hawaii.edu\"\n",
    "\n",
    "#this refers to the container (singularity/docker) used for the pipeline. These containers are located at /home/imibrahim/SIFs\n",
    "gatk_docker = \"arashi-gatk-426\"\n",
    "\n",
    "# read input csv\n",
    "input_df = pd.read_csv(input_file)\n",
    "#input must have\n",
    "# RG name, can be same as sample name\n",
    "# sample name\n",
    "# subject_id\n",
    "# sample type: T/N\n",
    "# absolute path to fastq1\n",
    "# absolute path to fastq2\n",
    "# library_name, can be same as sample name\n",
    "# platform unit, optional\n",
    "# sequence date\n",
    "# sequence platform\n",
    "# sequence center"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>readgroup</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>absolute_path_to_fq1</th>\n",
       "      <th>absolute_path_to_fq2</th>\n",
       "      <th>library_name</th>\n",
       "      <th>platform_unit</th>\n",
       "      <th>sequence_date</th>\n",
       "      <th>sequence_platform</th>\n",
       "      <th>sequence_center</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DCR002_NA</td>\n",
       "      <td>DCR002_NA</td>\n",
       "      <td>DCR002</td>\n",
       "      <td>N</td>\n",
       "      <td>/home/imibrahim/LTS/crc-temp/DCR002_NA/V350150...</td>\n",
       "      <td>/home/imibrahim/LTS/crc-temp/DCR002_NA/V350150...</td>\n",
       "      <td>DCR002_NA</td>\n",
       "      <td>DCR002_NA</td>\n",
       "      <td>2022-05-31T11:52:00</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>BGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DCR002_TA</td>\n",
       "      <td>DCR002_TA</td>\n",
       "      <td>DCR002</td>\n",
       "      <td>N</td>\n",
       "      <td>/home/imibrahim/LTS/crc-temp/DCR002_TA/V350150...</td>\n",
       "      <td>/home/imibrahim/LTS/crc-temp/DCR002_TA/V350150...</td>\n",
       "      <td>DCR002_TA</td>\n",
       "      <td>DCR002_TA</td>\n",
       "      <td>2022-05-31T11:52:00</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>BGI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   readgroup sample_name subject_id sample_type  \\\n",
       "0  DCR002_NA   DCR002_NA     DCR002           N   \n",
       "1  DCR002_TA   DCR002_TA     DCR002           N   \n",
       "\n",
       "                                absolute_path_to_fq1  \\\n",
       "0  /home/imibrahim/LTS/crc-temp/DCR002_NA/V350150...   \n",
       "1  /home/imibrahim/LTS/crc-temp/DCR002_TA/V350150...   \n",
       "\n",
       "                                absolute_path_to_fq2 library_name  \\\n",
       "0  /home/imibrahim/LTS/crc-temp/DCR002_NA/V350150...    DCR002_NA   \n",
       "1  /home/imibrahim/LTS/crc-temp/DCR002_TA/V350150...    DCR002_TA   \n",
       "\n",
       "  platform_unit        sequence_date sequence_platform sequence_center  \n",
       "0     DCR002_NA  2022-05-31T11:52:00          ILLUMINA             BGI  \n",
       "1     DCR002_TA  2022-05-31T11:52:00          ILLUMINA             BGI  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make sure to check if the interval file (`interval_loc`) used in the next block matches the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# switching of references depending on ref_version\n",
    "if ref_version == \"b37\":\n",
    "    interval_loc=f\"{ref_dir}/Agilent_SureSelect_Human_All_Exon_V4/S03723314_Regions.converted.bed\"\n",
    "\n",
    "    ref_fasta = f\"{ref_dir}/b37/human_g1k_v37_decoy.fasta\"\n",
    "    ref_fai = f\"{ref_dir}/b37/human_g1k_v37_decoy.fasta.fai\"\n",
    "    ref_dict = f\"{ref_dir}/b37/human_g1k_v37_decoy.dict\"\n",
    "    \n",
    "\n",
    "    pon = f\"{aligned_bam_dir}/pon/{groupname}_pon.vcf\"\n",
    "    pon_idx = f\"{aligned_bam_dir}/pon/{groupname}_pon.vcf.idx\"\n",
    "\n",
    "    gnomad = f\"{ref_dir}/b37/af-only-gnomad.raw.sites.vcf\"\n",
    "    gnomad_idx = f\"{ref_dir}/b37/af-only-gnomad.raw.sites.vcf.idx\"\n",
    "    variants_for_contamination = f\"{ref_dir}/b37/small_exac_common_3.vcf\"\n",
    "    variants_for_contamination_idx = f\"{ref_dir}/b37/small_exac_common_3.vcf.idx\"\n",
    "    funco_data_source = f\"{ref_dir}/b37/funcotator_dataSources.v1.7.20200521s.tar.gz\"\n",
    "elif ref_version == \"hg38\":\n",
    "    # only WGS version of interval list on disk for hg38\n",
    "    interval_loc = f\"{ref_dir}/hg38/wgs_calling_regions.hg38.interval_list\"\n",
    "\n",
    "    ref_fasta = f\"{ref_dir}/hg38/Homo_sapiens_assembly38.fasta\"\n",
    "    ref_fai = f\"{ref_dir}/hg38/Homo_sapiens_assembly38.fasta.fai\"\n",
    "    ref_dict = f\"{ref_dir}/hg38/Homo_sapiens_assembly38.dict\"\n",
    "\n",
    "    pon = f\"{{aligned_bam_dir}}/pon/1000g_pon.hg38.vcf.gz\"\n",
    "    pon_idx = f\"{{aligned_bam_dir}}/pon/1000g_pon.hg38.vcf.gz.tbi\"\n",
    "\n",
    "    gnomad = f\"{ref_dir}/hg38/af-only-gnomad.hg38.vcf.gz\"\n",
    "    gnomad_idx = f\"{ref_dir}/hg38/af-only-gnomad.hg38.vcf.gz.tbi\"\n",
    "    variants_for_contamination = f\"{ref_dir}/hg38/small_exac_common_3.hg38.vcf.gz\"\n",
    "    variants_for_contamination_idx = f\"{ref_dir}/hg38/small_exac_common_3.hg38.vcf.gz.tbi\"\n",
    "    funco_data_source = f\"{ref_dir}/hg38/funcotator_dataSources.v1.7.20200521s.tar.gz\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# A. SCMA input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 input files generated @\n",
      "/Users/isammohdibrahim/Documents/Dev/running-with-oliver-cromwell/output/scma\n"
     ]
    }
   ],
   "source": [
    "scma_output_dir=f'{base_output_dir}/scma'\n",
    "os.makedirs(scma_output_dir, exist_ok=True)\n",
    "\n",
    "for index, row in input_df.iterrows():\n",
    "    i = index + 1\n",
    "    if send_email_for_all == False and i == input_df.size[0]:\n",
    "        send_email = True\n",
    "    else:\n",
    "        send_email = send_email_for_all\n",
    "\n",
    "    output = {\n",
    "            \"seqConvMarkAdapt.readgroup_name\": row['readgroup'],\n",
    "            \"seqConvMarkAdapt.sample_name\": row['sample_name'],\n",
    "            \"seqConvMarkAdapt.fastq_1\": row['absolute_path_to_fq1'],\n",
    "            \"seqConvMarkAdapt.fastq_2\": row['absolute_path_to_fq2'],\n",
    "            \"seqConvMarkAdapt.library_name\": row['library_name'],\n",
    "            \"seqConvMarkAdapt.platform_unit\": row['platform_unit'],\n",
    "            \"seqConvMarkAdapt.run_date\": row['sequence_date'],\n",
    "            \"seqConvMarkAdapt.platform_name\": row['sequence_platform'],\n",
    "            \"seqConvMarkAdapt.sequencing_center\": row['sequence_center'],\n",
    "            \"seqConvMarkAdapt.make_fofn\": True,\n",
    "            \"seqConvMarkAdapt.send_email\": send_email,  \n",
    "            \"seqConvMarkAdapt.email\": email,  \n",
    "        }\n",
    "    \n",
    "    with open(f'{scma_output_dir}/{row[\"readgroup\"]}.json', 'w') as outfile:\n",
    "        json.dump(output, outfile, indent=4)\n",
    "    output = {}\n",
    "\n",
    "print(f'{i} input files generated @')\n",
    "print(f'{scma_output_dir}')\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# B. Pre-processing\n",
    "- no input needs to be generated. Pipeline will use output from SCMA."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# C. PoN input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pon_output_dir = f'{base_output_dir}/m2'\n",
    "os.makedirs(pon_output_dir, exist_ok=True)\n",
    "\n",
    "n = input_df.query('sample_type == \"N\"')\n",
    "samples_n = n['sample_name'].sort_values().unique().tolist()\n",
    "n_bams = [f'{n_dir}/{n}.{ref_version}.bam' for n in samples_n]\n",
    "n_bais = [f'{n_dir}/{n}.{ref_version}.bai' for n in samples_n]\n",
    "\n",
    "input = {\n",
    "    \"Mutect2_Panel.gatk_docker\": gatk_docker,\n",
    "\n",
    "    \"Mutect2_Panel.pon_name\": f\"{groupname}_pon\",\n",
    "    \"Mutect2_Panel.normal_bams\": n_bams,\n",
    "    \"Mutect2_Panel.normal_bais\": n_bais,\n",
    "\n",
    "\n",
    "    \"Mutect2_Panel.ref_fasta\": ref_fasta,\n",
    "    \"Mutect2_Panel.ref_fai\": ref_fai,\n",
    "    \"Mutect2_Panel.ref_dict\": ref_dict,\n",
    "    \"Mutect2_Panel.scatter_count\": 1,\n",
    "    \n",
    "    \"Mutect2_Panel.gnomad\": gnomad,\n",
    "    \"Mutect2_Panel.gnomad_idx\": gnomad_idx,\n",
    "\n",
    "    \"Mutect2_Panel.intervals\":interval_loc,\n",
    "    \"Mutect2_Panel.email\": email,\n",
    "    \"Mutect2_Panel.Mutect2.filter_mem\": 2000,\n",
    "    \"Mutect2_Panel.send_email\": True if send_email_for_all else False\n",
    "}\n",
    "\n",
    "with open(f'{pon_output_dir}/pon.json', 'w') as f:\n",
    "    json.dump(input, f, indent=4)\n",
    "\n",
    "print('PoN input generated at:')\n",
    "print(f'{pon_output_dir}/pon.json')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# D. Mutect2 input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "m2_output_dir = f'{base_output_dir}/m2'\n",
    "\n",
    "i = 0\n",
    "os.makedirs(m2_output_dir, exist_ok=True)\n",
    "\n",
    "# get list of subject ids\n",
    "subject_ids = input_df[\"subject_id\"].unique()\n",
    "t = input_df.query('sample_type == \"T\"')\n",
    "\n",
    "# parse through list and get tumor and normal sample id for each one\n",
    "for s in subject_ids:\n",
    "    i+= 1\n",
    "    \n",
    "    # this just attaches a send email job upon completion of the last workflow in the group\n",
    "    # send_email = True if i == len(subject_ids) else False\n",
    "    send_email = True\n",
    "    \n",
    "    n_sample = n.query(f'subject_id == \"{s}\" ')[\"sample_name\"].values[0]\n",
    "    t_sample = t.query(f'subject_id == \"{s}\" ')[\"sample_name\"].values[0]\n",
    "    m2_input = {\n",
    "        \n",
    "        \"Mutect2.normal_reads\": f\"{n_dir}/{n_sample}.{ref_version}.bam\",\n",
    "        \"Mutect2.normal_reads_index\": f\"{n_dir}/{n_sample}.{ref_version}.bai\",\n",
    "        \"Mutect2.tumor_reads\": f\"{t_dir}/{t_sample}.{ref_version}.bam\",\n",
    "        \"Mutect2.tumor_reads_index\": f\"{t_dir}/{t_sample}.{ref_version}.bai\",\n",
    "        \n",
    "        \"Mutect2.gatk_docker\": gatk_docker,\n",
    "  \n",
    "        \"Mutect2.intervals\": interval_loc,\n",
    "        \"Mutect2.scatter_count\": 24,\n",
    "        \"Mutect2.m2_extra_args\": \" -ip 100 \",\n",
    "        \"Mutect2.split_intervals_extra_args\": \" --subdivision-mode BALANCING_WITHOUT_INTERVAL_SUBDIVISION --dont-mix-contigs --min-contig-size 1000000 \",\n",
    "\n",
    "        \"Mutect2.filter_funcotations\": \"True\",\n",
    "        \"Mutect2.funco_reference_version\": ref_version, \n",
    "        # funcotator sources for all references are in here\n",
    "        \"Mutect2.funco_data_sources_tar_gz\": funco_data_source,\n",
    "\n",
    "        \"Mutect2.ref_fasta\": ref_fasta, \n",
    "        \"Mutect2.ref_fai\": ref_fai, \n",
    "        \"Mutect2.ref_dict\": ref_dict, \n",
    "\n",
    "        \"Mutect2.pon\": pon, \n",
    "        \"Mutect2.pon_idx\": pon_idx, \n",
    "\n",
    "        \"Mutect2.gnomad\": gnomad, \n",
    "        \"Mutect2.gnomad_idx\": gnomad_idx, \n",
    "        \"Mutect2.variants_for_contamination\": variants_for_contamination, \n",
    "        \"Mutect2.variants_for_contamination_idx\": variants_for_contamination_idx, \n",
    "\n",
    "        \"Mutect2.run_funcotator\": True,\n",
    "        \"Mutect2.run_orientation_bias_mixture_model_filter\": True,\n",
    "        \"Mutect2.send_email\": send_email\n",
    "    }\n",
    "    with open(f'{m2_output_dir}/{s}.json', 'w') as f:\n",
    "        json.dump(m2_input, f, indent=4)\n",
    "print(f'{i} inputs generated @:')\n",
    "print(f'{m2_output_dir}')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
